core/camel-xml-jaxb/src/main/java/org/apache/camel/xml/jaxb/JaxbXmlRoutesBuilderLoader.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.xml.jaxb;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.CamelContextAware;
+import org.apache.camel.RoutesBuilder;
+import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.model.RouteTemplatesDefinition;
+import org.apache.camel.model.RoutesDefinition;
+import org.apache.camel.model.rest.RestsDefinition;
+import org.apache.camel.spi.Resource;
+import org.apache.camel.spi.RoutesBuilderLoader;
+
+import static org.apache.camel.xml.jaxb.JaxbHelper.loadRestsDefinition;
+import static org.apache.camel.xml.jaxb.JaxbHelper.loadRouteTemplatesDefinition;
+import static org.apache.camel.xml.jaxb.JaxbHelper.loadRoutesDefinition;
+
+public class JaxbXmlRoutesBuilderLoader implements RoutesBuilderLoader, CamelContextAware {
```

Comment:
We should use the `@JdkService` annotation as then the maven tooling generated all the needed files for META-INF and the license files and so on, in the src/generated folder.

See for the how we use
@JdkService(ModelToXMLDumper.FACTORY)

Reply:
does this support directories ? i.e. `@JdkService("loaders/xml")`

Reply:
Oh this is just to find the factory itself (eg look in META-INF/services/..../, eg what you do at RoutesBuilderLoader.LOADERS_FACTORY_PATH
Its not for loading resources.

Reply:
So the `RoutesBuilderLoader` is not actually discovered using a service mechanism but in fact the default implementation is a facade to the real one which behind the scene id does:

```java
    @Override
    public RoutesBuilder loadRoutesBuilder(Resource resource) throws Exception {
        // language is derived from the file extension
        final String language = FileUtil.onlyExt(resource.getLocation(), true);

        return getLoader(language).loadRoutesBuilder(resource);
    }

    private RoutesBuilderLoader getLoader(String language) {
        RoutesBuilderLoader answer = getCamelContext().getRegistry().lookupByNameAndType(
            language, 
            RoutesBuilderLoader.class);

        if (answer == null) {
            // lookup the loader in
            // META-INF/services/org/apache/camel/routes-loaders/${language}
        }

        return answer;
    }
```

So it looks up the real loader using the file extension, and the real loader is supposed to be either bound to the registry or defined through `META-INF/services/org/apache/camel/routes-loaders`. I'm not very convinced by this patter so I'll split the loader interface from the language specific one.

Reply:
Ah okay yeah that is another layer. But that seems fine to do. You may want to throw some kind of appropiate exceprion if a loader cannot be found for a given language.

Reply:
So I split the implementation in two interfaces:

- **RoutesBuilderLoader** which is in charge to load a resource into a `RoutesBuilder`
- **RoutesLoader** which is in charge to add the routes to the camel context. Behind the scenes it uses the registry and the factory finder to find the appropriate `RoutesBuilderLoader` for a given resource based on the file extension.